src/App.tsx

```diff
@@ -1,24 +1,217 @@
+        setTodos(todosFromServer);
+      })
+      .catch(() => setError('Error: cannot upload todos'));
+  }, []);
+
+  useEffect(() => {
+    let errorTimer: number;
+    let hiddenTime: number;
+
+    if (error) {
+      hiddenTime = window.setTimeout(() => {
+        setIsHidden(true);
+      }, 2000);
+      errorTimer = window.setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => {
+      clearTimeout(errorTimer);
+      clearTimeout(hiddenTime);
+    };
+  }, [error]);
+
+  const completedTodos = todos.filter((todo) => todo.completed);
+  const activeTodos = todos.filter((todo) => !todo.completed);
+
+  const visibleTodos: Todo[] = useMemo(() => {
+    switch (filterMethod) {
+      case FilterOptions.ACTIVE:
+        return activeTodos;
+
+      case FilterOptions.COMPLETED:
+        return completedTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, filterMethod]);
+
+  const addTodo = async (title: string) => {
+    try {
+      const newTodo = {
+        title,
+        userId: USER_ID,
+        completed: false,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add todo');
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const deleteTodo = async (todoId: number) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoId]);
+      await removeTodo(todoId);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete todo');
+    } finally {
+      setLoadingTodos(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  };
+
+  const deleteComplededTodos = async () => {
+    const deletePromises = completedTodos.map(todo => deleteTodo(todo.id));
+
+    try {
+      await Promise.all(deletePromises);
+    } catch {
+      setError('Unable to delete todos');
+    }
+  };
+
+  const toggleTodo = async (
+    todoId: number,
+    args: UpdateTodoArgs,
+  ) => {
+    try {
+      setLoadingTodos((prevState) => [...prevState, todoId]);
+
+      await updateTodo(todoId, args);
+
+      setTodos((prevState) => prevState.map(todo => {
+        return (todo.id === todoId)
+          ? {
+            ...todo,
+            completed: !todo.completed,
+          }
+          : todo;
+      }));
+    } catch (err) {
+      setError(`${err}`);
+    } finally {
+      setLoadingTodos(
+        (prevState) => prevState.filter((id => id !== todoId)),
+      );
+    }
+  };
+
+  const toggleAllTodos = async () => {
+    const hasUncompletedTodos = todos.some((todo) => !todo.completed);
+
+    let actionTodos: Todo[] = [];
+
+    actionTodos = !hasUncompletedTodos
+      ? completedTodos
+      : activeTodos;
+
+    actionTodos.forEach(async todoToUpdate => {
+      await toggleTodo(todoToUpdate.id,
```

Comment:
I noticed that you use 2 different approaches in the project `async/await` and `then` for working with async functions. It's  generally recommended to be consistent and use a single approach throughout a project

Reply:
good advice, thank you. I changed everything approaches in the project to async/await except one file. It was a template from master branch and I think that it is better not to change. Please tell me if it is better to change even in template.

Reply:
Yeah, you can leave it like that